### PR TITLE
[core][gcs] Make GCS client working with timeout_ms.

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?
In [PR](https://github.com/ray-project/ray/pull/24764) we move the reconnection to GcsRPCClient. In case of a GCS failure, we'll queue the requests and resent them once GCS is back.
This actually breaks request with timeout because  now, the request will be queued and never got a response. This PR fixed it.

For all requests, it'll be stored by the time it's supposed to be timeout. When GCS is down, we'll check the queued requests and make sure if it's timeout, we'll reply immediately with a Timeout error message.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/rpc/gcs_server/gcs_rpc_client.h [hunk 1]
```diff
@@ -32,21 +33,26 @@ class GcsRpcClient;
 /// Executor saves operation and support retries.
 class Executor {
  public:
-  explicit Executor(GcsRpcClient *gcs_rpc_client) : gcs_rpc_client_(gcs_rpc_client) {}
+  explicit Executor(GcsRpcClient *gcs_rpc_client,
```
Comment (mwtian) [inline src/ray/rpc/gcs_server/gcs_rpc_client.h:36]: `explicit` should be unnecessary. Also, take `const ray::Status &` in the `abort_callback`?

File: src/ray/rpc/gcs_server/gcs_rpc_client.h [hunk 2]
```diff
@@ -32,21 +33,26 @@ class GcsRpcClient;
...
+                    std::function<void(ray::Status)> abort_callback)
+      : gcs_rpc_client_(gcs_rpc_client), abort_callback_(std::move(abort_callback)) {}
 
   /// This function is used to execute the given operation.
   ///
   /// \param operation The operation to be executed.
-  void Execute(const std::function<void(GcsRpcClient *gcs_rpc_client)> &operation) {
+  void Execute(std::function<void(GcsRpcClient *gcs_rpc_client)> operation) {
     operation_ = operation;
```
Comment (mwtian) [inline src/ray/rpc/gcs_server/gcs_rpc_client.h:44]: `std::move()`?

File: src/ray/rpc/gcs_server/gcs_rpc_client.h [hunk 3]
```diff
@@ -32,21 +33,26 @@ class GcsRpcClient;
...
     operation(gcs_rpc_client_);
   }
 
   /// This function is used to retry the given operation.
   void Retry() { operation_(gcs_rpc_client_); }
 
+  void Abort(ray::Status status) { abort_callback_(status); }
```
Comment (mwtian) [inline src/ray/rpc/gcs_server/gcs_rpc_client.h:51]: `const ray::Status &`?

File: src/ray/rpc/gcs_server/gcs_rpc_client.h [hunk 4]
```diff
@@ -496,10 +506,27 @@ class GcsRpcClient {
...
+    // We need to cleanup all the pending requets which are timeout.
+    auto now = absl::Now();
+    while (!pending_requests_.empty()) {
+      auto iter = pending_requests_.begin();
+      if (iter->first > now) {
+        break;
+      }
+      auto [executor, request_bytes] = iter->second;
+      executor->Abort(
+          ray::Status::TimedOut("Time out due to GCS is down. Abort this operation"));
```
Comment (mwtian) [inline src/ray/rpc/gcs_server/gcs_rpc_client.h:524]: Maybe "Timed out while waiting for GCS to become available."

**Review by mwtian — APPROVED** [review]
Btw I think if the retry happens inside gRPC client, timeout should be handled as well.

**Comment by iycheng** [comment]
@mwtian I think if it timeout inside gRPC, this will be handled well even without this PR. This PR will only cover the case where it failed not because of timeout but because of some other issues (like unavailable)

https://github.com/grpc/proposal/blob/master/A6-client-retries.md

gRPC retry only support retry for 5 times, so we can't depend on that one to handle the retry for us :(

**Pull request merged**